uaclient/http/__init__.py

```diff
@@ -0,0 +1,199 @@
+    for env_var in ("no_proxy", "NO_PROXY"):
+        proxy_value = os.environ.get(env_var)
+        if proxy_value:
+            # Honor no proxy values and extend UA-specific where absent
+            no_proxy = ",".join(
+                sorted(
+                    set(proxy_value.split(",")).union(set(UA_NO_PROXY_URLS))
+                )
+            )
+    logging.debug("Setting no_proxy: %s", no_proxy)
+    os.environ["no_proxy"] = no_proxy
+    os.environ["NO_PROXY"] = no_proxy
+    if proxy_dict:
+        proxy_handler = request.ProxyHandler(proxy_dict)
+        opener = request.build_opener(proxy_handler)
+        request.install_opener(opener)
+
+
+def _readurl_urllib(
+    req: request.Request, timeout: Optional[int] = None
+) -> UnparsedHTTPResponse:
+    try:
+        resp = request.urlopen(req, timeout=timeout)
+    except error.HTTPError as e:
+        resp = e
+    except error.URLError as e:
+        raise exceptions.UrlError(e, url=req.full_url)
+
+    body = resp.read().decode("utf-8")
+
+    # convert EmailMessage header object to dict with lowercase keys
+    headers = {k.lower(): v for k, v, in resp.headers.items()}
+
+    return UnparsedHTTPResponse(
+        code=resp.code,
+        headers=headers,
+        body=body,
+    )
+
+
+def readurl(
+    url: str,
+    data: Optional[bytes] = None,
+    headers: Dict[str, str] = {},
+    method: Optional[str] = None,
+    timeout: Optional[int] = None,
+    log_response_body: bool = True,
+) -> HTTPResponse:
+    if data and not method:
+        method = "POST"
+    req = request.Request(url, data=data, headers=headers, method=method)
+
+    sorted_header_str = ", ".join(
+        ["'{}': '{}'".format(k, headers[k]) for k in sorted(headers)]
+    )
+    logging.debug(
+        "URL [{}]: {}, headers: {{{}}}, data: {}".format(
+            method or "GET",
+            url,
+            sorted_header_str,
+            data.decode("utf-8") if data else None,
+        )
+    )
+
+    resp = _readurl_urllib(req, timeout=timeout)
+
+    json_dict = {}
+    json_list = []
+    if "application/json" in resp.headers.get("content-type", ""):
+        json_body = json.loads(resp.body, cls=util.DatetimeAwareJSONDecoder)
+        if isinstance(json_body, dict):
+            json_dict = json_body
+        elif isinstance(json_body, list):
+            json_list = json_body
+
+    sorted_header_str = ", ".join(
+        ["'{}': '{}'".format(k, resp.headers[k]) for k in sorted(resp.headers)]
+    )
+    debug_msg = "URL [{}] response: {}, headers: {{{}}}".format(
+        method or "GET", url, sorted_header_str
+    )
+    if log_response_body:
+        # Due to implicit logging redaction, large responses might take longer
+        body_to_log = resp.body  # type: Any
+        if json_dict:
+            body_to_log = json_dict
+        elif json_list:
+            body_to_log = json_list
+        debug_msg += ", data: {}".format(body_to_log)
```

Comment:
I know this is a weird scenario and I am not sure if we want to cover it, but should we log both items if they exist ?

Reply:
I'd rather log both of them in the "extra" fields of the log, but that would be a change to the log format and would require changes to how we redact things from these logs. So I thought that this PR should keep the exact same log format as before, which I think is what this does. What do you think?

Reply:
That's fair. We can change that on a following PR if we think it would be useful in the future